node/core/av-store/src/lib.rs

```diff
@@ -110,62 +348,201 @@ impl AvailabilityStoreSubsystem {
 			inner: Arc::new(db),
 			metrics,
 		})
 	}
 
 	#[cfg(test)]
-	fn new_in_memory(inner: Arc<dyn KeyValueDB>) -> Self {
+	fn new_in_memory(inner: Arc<dyn KeyValueDB>, pruning_config: PruningConfig) -> Self {
 		Self {
+			pruning_config,
 			inner,
 			metrics: Metrics(None),
 		}
 	}
 }
 
-async fn run<Context>(subsystem: AvailabilityStoreSubsystem, mut ctx: Context)
+fn get_next_pov_pruning_time(db: &Arc<dyn KeyValueDB>) -> Option<NextPoVPruning> {
+	query_inner(db, columns::META, &NEXT_POV_PRUNING)
+}
+
+fn get_next_chunk_pruning_time(db: &Arc<dyn KeyValueDB>) -> Option<NextChunkPruning> {
+	query_inner(db, columns::META, &NEXT_CHUNK_PRUNING)
+}
+
+async fn run<Context>(mut subsystem: AvailabilityStoreSubsystem, mut ctx: Context)
 	-> Result<(), Error>
 where
 	Context: SubsystemContext<Message=AvailabilityStoreMessage>,
 {
 	let ctx = &mut ctx;
 	loop {
+		let pov_pruning_time = subsystem.maybe_prune_povs()?;
+		let chunk_pruning_time = subsystem.maybe_prune_chunks()?;
+
+		let mut pov_pruning_time = pov_pruning_time.fuse();
+		let mut chunk_pruning_time = chunk_pruning_time.fuse();
+
 		select! {
 			incoming = ctx.recv().fuse() => {
 				match incoming {
-					Ok(FromOverseer::Signal(Conclude)) => break,
-					Ok(FromOverseer::Signal(_)) => (),
+					Ok(FromOverseer::Signal(OverseerSignal::Conclude)) => break,
+					Ok(FromOverseer::Signal(OverseerSignal::ActiveLeaves(
+						ActiveLeavesUpdate { activated, .. })
+					)) => {
+						for activated in activated.into_iter() {
+							process_block_activated(ctx, &subsystem.inner, activated).await?;
+						}
+					}
+					Ok(FromOverseer::Signal(OverseerSignal::BlockFinalized(hash))) => {
+						process_block_finalized(&subsystem, ctx, &subsystem.inner, hash).await?;
+					}
 					Ok(FromOverseer::Communication { msg }) => {
-						process_message(&subsystem.inner, &subsystem.metrics, msg)?;
+						process_message(&mut subsystem, ctx, msg).await?;
 					}
 					Err(_) => break,
 				}
 			}
+			pov_pruning_time = pov_pruning_time => {
+				subsystem.prune_povs()?;
+			}
+			chunk_pruning_time = chunk_pruning_time => {
+				subsystem.prune_chunks()?;
+			}
 			complete => break,
 		}
 	}
 
 	Ok(())
 }
 
-fn process_message(db: &Arc<dyn KeyValueDB>, metrics: &Metrics, msg: AvailabilityStoreMessage) -> Result<(), Error> {
+/// As soon as certain block is finalized its pruning records and records of all
+/// blocks that we keep that are `older` than the block in question have to be updated.
+///
+/// The state of data has to be changed from
+/// `CandidateState::Included` to `CandidateState::Finalized` and their pruning times have
+/// to be updated to `now` + keep_finalized_{block, chunk}_for`.
+async fn process_block_finalized<Context>(
+	subsystem: &AvailabilityStoreSubsystem,
+	ctx: &mut Context,
+	db: &Arc<dyn KeyValueDB>,
+	hash: Hash,
+) -> Result<(), Error>
+where
+	Context: SubsystemContext<Message=AvailabilityStoreMessage>
+{
+	let block_number = get_block_number(ctx, hash).await?;
```

Comment:
can we rename this function / fix the parameter names? Because it certainly doesn't return the block number of the provided hash.

Reply:
I don't think I understand

Reply:
maybe @pepyakin is referring to the fact that `get_block_number` does `map(|number| number + 1)` which is not obvious given the name of the function?

Reply:
That's right.

Looking only at the call site, a reader would most likely think that this function returns a block number of a given block hash, whereas the function actually returns the block number following the given hash.

Then, the `block_number` name is also a bit confusing, since it is a block number following the hash `hash` which is the latest finalized block. So if not better, but definitely a more honest name would be, `finilized_block_number_plus_1`.

I had hoped that during renaming that would be also clarified/fixed.